src/Nonlinear/ReverseAD/reverse_mode.jl

```diff
@@ -0,0 +1,265 @@
+        _reverse_eval(con)
+    end
+    copyto!(d.last_x, x)
+    return
+end
+
+"""
+    _forward_eval(
+        f::Union{_FunctionStorage,_SubexpressionStorage},
+        d::NLPEvaluator,
+        x::AbstractVector{T},
+    ) where {T}
+
+Forward-mode evaluation of an expression tree given in `f`.
+
+ * This function assumes that the values of all sub-expressions have already
+   been computed and are stored in `d.subexpression_forward_values`.
+ * `f.partials_storage[k]` is the partial derivative of `nodes[k].parent` with
+   respect to the value of node `k`. It's efficient to compute this at the same
+   time as the value of the parent because we use it in reverse mode and in dual
+   forward mode. Note that `partials_storage`` makes a subtle assumption that we
+   have a tree instead of a general DAG. If we have a DAG, then need to
+   associate storage with each edge of the DAG.
+"""
+function _forward_eval(
+    # !!! warning
+    #     This Union depends upon _FunctionStorage and _SubexpressionStorage
+    #     having similarly named fields.
+    f::Union{_FunctionStorage,_SubexpressionStorage},
+    d::NLPEvaluator,
+    x::AbstractVector{T},
+)::T where {T}
+    @assert length(f.forward_storage) >= length(f.nodes)
+    @assert length(f.partials_storage) >= length(f.nodes)
+    operators = d.data.operators
+    # f.nodes is already in order such that parents always appear before
+    # children, so a backwards pass through f.nodes is a forward pass through
+    # the tree.
+    children_arr = SparseArrays.rowvals(f.adj)
+    for k in length(f.nodes):-1:1
+        node = f.nodes[k]
+        f.partials_storage[k] = zero(T)
+        if node.type == Nonlinear.NODE_VARIABLE
+            f.forward_storage[k] = x[node.index]
+            # This should never happen, because we will have replaced these by now.
+            # elseif node.type == Nonlinear.NODE_MOI_VARIABLE
+            #     f.forward_storage[k] = x[node.index]
+        elseif node.type == Nonlinear.NODE_VALUE
+            f.forward_storage[k] = f.const_values[node.index]
+        elseif node.type == Nonlinear.NODE_SUBEXPRESSION
+            f.forward_storage[k] = d.subexpression_forward_values[node.index]
+        elseif node.type == Nonlinear.NODE_PARAMETER
+            f.forward_storage[k] = d.data.parameters[node.index]
+        elseif node.type == Nonlinear.NODE_CALL_MULTIVARIATE
+            children_indices = SparseArrays.nzrange(f.adj, k)
+            N = length(children_indices)
+            f_input = _UnsafeVectorView(d.jac_storage, N)
+            ∇f = _UnsafeVectorView(d.user_output_buffer, N)
+            for (r, i) in enumerate(children_indices)
+                f_input[r] = f.forward_storage[children_arr[i]]
+                ∇f[r] = 0.0
+            end
+            f.forward_storage[k] = Nonlinear.eval_multivariate_function(
+                operators,
+                operators.multivariate_operators[node.index],
+                f_input,
+            )
```

Comment:
I've only made one change compared to the implementation in JuMP, and that is to push all multivariate functions through `eval_multivariate_function`, rather than hard-coded things here: https://github.com/jump-dev/JuMP.jl/blob/3ddee2fd1ac2d8d304130073ed09e051a599582b/src/_Derivatives/forward.jl#L70-L161

I think we should favor merging this as-is, because it's a lot more readable. We can always revert the change later if benchmarking shows that it is a problem.

Reply:
I've re-added the old JuMP code for evaluating multivariate functions. It's simple to remove with benchmarking in future.